File: lox/parser/token.cpp
#include <lox/parser/token.hpp>

#include <fmt/std.h>

#include <utility>

namespace lox {

constexpr std::array token_type_names{
    "LEFT_PAREN", "RIGHT_PAREN", "LEFT_BRACE", "RIGHT_BRACE", "COMMA", "DOT",
    "MINUS", "PLUS", "SEMICOLON", "SLASH", "STAR", "QUESTION", "COLON",

    // One or two character tokens
    "BANG", "BANG_EQUAL", "EQUAL", "EQUAL_EQUAL", "GREATER", "GREATER_EQUAL",
    "LESS", "LESS_EQUAL",

    // Literals
    "IDENTIFIER", "STRING", "NUMBER",

    // Keywords
    "AND", "CLASS", "ELSE", "FALSE", "FUN", "FOR", "IF", "NIL", "OR", "PRINT",
    "RETURN", "SUPER", "THIS", "TRUE", "VAR", "WHILE", "BREAK",

    "EOF"};

// Ensure that all token types can be looked up
static_assert(size(token_type_names) ==
              static_cast<std::size_t>(token_type::NUM_TYPES));

constexpr auto to_string(token_type type) -> const char* {
  return token_type_names.at(static_cast<std::size_t>(type));
}

// clang-format off
template <class... Ts> struct overloaded : Ts... { using Ts::operator()...; }; 
template <class... Ts> overloaded(Ts...) -> overloaded<Ts...>;
// clang-format on

auto is_truthy(literal literal) -> bool {
  return std::visit(overloaded{[](std::monostate&) { return false; },
                               [](bool arg) { return arg; },
                               [](auto&&) { return true; }},
                    literal);
}

auto to_string(literal literal) -> std::string {
  using namespace std::string_literals;
  return std::visit(
      overloaded{[](std::monostate) { return "nil"s; },
                 [](bool arg) { return arg ? "true"s : "false"s; },
                 [](double arg) { return fmt::format("{}", arg); },
                 [](std::string arg) { return fmt::format("\"{}\"", arg); }},
      literal);
}

auto to_string(token token) -> std::string {
  return token.literal.index() == 0U
           ? fmt::format("{}", to_string(token.type))
           : fmt::format("{} {} {}", to_string(token.type), token.lexeme,
                         token.literal);
}

} // namespace lox

File: glox/internal/ltoken/token.go
package ltoken

import "fmt"

//go:generate stringer -type=Type
type Type int

const (
	// Single-character tokens
	LeftParen Type = iota
	RightParen
	LeftBrace
	RightBrace
	Comma
	Dot
	Minus
	Plus
	Semicolon
	Slash
	Star

	// One or two character tokens
	Bang
	BangEqual
	Equal
	EqualEqual
	Greater
	GreaterEqual
	Less
	LessEqual

	// Literals
	Identifier
	String
	Number

	// Keywords
	And
	Class
	Else
	False
	Fun
	For
	If
	Nil
	Or
	Print
	Return
	Super
	This
	True
	Var
	While

	EOF
)

var keywords = map[string]Type{
	"and":    And,
	"class":  Class,
	"else":   Else,
	"false":  False,
	"for":    For,
	"fun":    Fun,
	"if":     If,
	"nil":    Nil,
	"or":     Or,
	"print":  Print,
	"return": Return,
	"super":  Super,
	"this":   This,
	"true":   True,
	"var":    Var,
	"while":  While,
}

func LookupKeyword(identifier string) Type {
	ttype, ok := keywords[identifier]
	if !ok {
		return Identifier
	}
	return ttype
}

type Token struct {
	Type    Type
	Lexeme  string
	Literal any // TODO: How to represent this?
	Line    int
}

func (t Token) String() string {
	switch t.Type {
	case EOF:
		return "EOF"
	case String:
		return fmt.Sprintf("%s(%q)", t.Type, t.Literal)
	case Identifier, Number:
		return fmt.Sprintf("%s(%v)", t.Type, t.Literal)
	default:
		return fmt.Sprintf("%s", t.Type)
	}
}
